pytorch_multi_things_2.py

```diff
@@ -0,0 +1,221 @@
+
+    def train_batch(self, batch):
+        input, target = batch
+
+        # I propose we handle zero_grad() automatically... is there a downside to this?
+        # self.optimizer.zero_grad()
+
+        with autocast():
+            output = self.model(input)
+            loss = loss_fn(output, target)
+
+        self.context.scale(loss).backward()
+
+        # Since we are going to support gradient aggregation internally, we might ignore this call.
+        self.context.step(self.optimizer)
+
+        # No need for the user to call scaler.update(); we'll handle that for them.
+
+
+"""
+Example 2: multiple objects with new pytorch API
+
+https://pytorch.org/docs/master/notes/amp_examples.html#working-with-multiple-models-losses-and-optimizers
+"""
+
+scaler = torch.cuda.amp.GradScaler()
+
+for epoch in epochs:
+    for input, target in data:
+        optimizer0.zero_grad()
+        optimizer1.zero_grad()
+        with autocast():
+            output0 = model0(input)
+            output1 = model1(input)
+            loss0 = loss_fn(2 * output0 + 3 * output1, target)
+            loss1 = loss_fn(3 * output0 - 5 * output1, target)
+
+        scaler.scale(loss0).backward(retain_graph=True)
+        scaler.scale(loss1).backward()
+
+        # You can choose which optimizers receive explicit unscaling, if you
+        # want to inspect or modify the gradients of the params they own.
+        scaler.unscale_(optimizer0)
+
+        scaler.step(optimizer0)
+        scaler.step(optimizer1)
+
+        scaler.update()
+
+
+# The same thing in Determined:
+
+class MyPytorchTrial(PyTorchTrial):
+    def __init__(self, context):
+        self.context = context
+        self.scaler = self.context.set_scaler(GradScaler())
+        self.model0 = self.context.model(my_make_model())
+        self.model1 = self.context.model(my_make_model())
+        self.optimizer0 = self.context.optimizer(my_make_optimizer(self.model0.parameters()))
+        self.optimizer1 = self.context.optimizer(my_make_optimizer(self.model1.parameters()))
+
+    def train_batch(self, batch):
+        input, target = batch
+
+        # We deal with zero_grad() (do we?)
+
+        with autocast():
+            output0 = self.model0(input)
+            output1 = self.model1(input)
+            loss0 = loss_fn(2 * output0 + 3 * output1, target)
+            loss1 = loss_fn(3 * output0 - 5 * output1, target)
+
+        self.scaler.scale(loss0).backward(retain_graph=True)
+        self.scaler.scale(loss1).backward()
+
+        # Users clip gradients in callbacks specified in the scaler.step() operation
+        # (this is because we handle gradient aggregation ourselves)
+        def my_clip_grads_0(...):
+            return ...
+
+        def my_clip_grads_1(...):
+            return ...
+
+        self.scaler.step(self.optimizer0, clip_grads=my_clip_grads_0)
+        self.scaler.step(self.optimizer1, clip_grads=my_clip_grads_1)
```

Comment:
I like the idea of it being similar to the pytorch-native version. It allows for easier porting, and allows the user the controls they want. I think we could handle the optim.zero_grad ourselves especially since we handle gradient aggregation.

Reply:
> I think we could handle the optim.zero_grad ourselves especially since we handle gradient aggregation

Cool!  That is also what @aaron276h said, so if you both feel that way I feel extra good about this.